### DISCUSSION TITLE
Idiomatic use of smart pointers in cpp2 code

### DISCUSSION THREAD

**Discussion by AbhinavK00** [hsutter/cppfront#674] [Q&A]
I've never worked on large codebases so apologies in advance if this question turns out to be stupid. 
The question, how to use smart pointers in cpp2 code.
Suppose I have a function which takes a `bigStruct` as inout parameter. It works with stack values but if I want to handle the case for a value on free store, what should I do? My first instinct was to write another overload of function which takes a unique pointer to bigStruct. But even with this, should that function take the unique pointer as an `inout` parameter or a `move` parameter?
If the former, wouldn't that require the `get` member to be called?
If the latter, how do I garuntee to the caller that I have not stolen the guts to his resource (since my intention was `inout`)?
If any of those, what should I expect the caller side code to look like?
Or are there other ways to do this?

Notes: bigStruct is any expensive to copy thing

**Comment by JohelEGP** [comment]
Same as Cpp1.
See [F.7: For general use, take `T*` or `T&` arguments rather than smart pointers](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-smart).
See its "see also".
